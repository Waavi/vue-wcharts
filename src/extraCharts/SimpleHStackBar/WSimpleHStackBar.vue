
<template>
    <div class="WSimpleHStackBar">
        <div
            v-if="stacks"
            class="Container"
        >
            <div
                v-if="launchAnimation"
                class="Start"
            >
                <slot
                    name="start"
                    :total="sumValues"
                />
            </div>

            <div class="Stacks">
                <div
                    v-for="(stack, index) in stacks"
                    v-show="!(stack.value <= 0)"
                    :id="stack.id"
                    :key="`stack-${stack.id}`"
                    class="Stack"
                    :class="{ First: index === 0, Last: index === stacks.length - 1 }"
                    :style="{
                        ...barStylesCmp,
                        opacity: launchAnimation ? 1 : 0,
                        background: stack.color,
                        maxWidth: `${stack.width}%`,
                    }"
                    v-on="stackListeners"
                >
                    <div
                        class="Value"
                        :style="labelStylesCmp"
                    >
                        <WShowIfFit>
                            <slot
                                v-if="!stack.hide && showLabel"
                                name="value"
                                :index="stack.id"
                                :value="stack.value"
                                :percentage="stack.width|percentage"
                                :color="stack.color"
                            >
                                <span
                                    :style="{
                                        color: stack.color
                                    }"
                                >{{ stack.value }}</span>
                            </slot>
                        </WShowIfFit>
                    </div>
                </div>
            </div>

            <WTooltip>
                <template #default="{ value: [data] }">
                    <div class="Toolip">
                        <WBullet :styles="{ background: data.color }" />
                        <slot
                            name="tooltip"
                            v-bind="data"
                        >
                            {{ data.value }}
                        </slot>
                    </div>
                </template>
            </WTooltip>

            <div
                v-for="marker in stackMarkers"
                v-show="!(marker.value <= 0)"
                :id="marker.id"
                :key="`marker-${marker.id}`"
                data-type="marker"
                class="Marker"
                :style="{
                    left: `${marker.left}%`,
                    ...markerStylesCmp,
                }"
                v-on="markerListeners"
            >
                <slot
                    name="marker"
                    :value="marker"
                >
                    <div :data-value="marker.value" />
                </slot>
            </div>

            <div
                v-if="launchAnimation"
                class="End"
            >
                <slot
                    name="end"
                    :total="sumValues"
                />
            </div>
        </div>
    </div>
</template>

<script>
import VueTypes from 'vue-types'
import merge from '../../utils/merge'
import { WTooltip, WBullet, WShowIfFit } from '../../components/Widgets'
import activeMixin from '../../mixins/active'
import animationMixin from '../../mixins/animation'
import themeMixin from '../../mixins/theme'

const MIN_WIDTH = 4

export default {
    name: 'WSimpleHStackBar',
    components: {
        WTooltip,
        WBullet,
        WShowIfFit,
    },
    provide () {
        return {
            Chart: this,
        }
    },
    filters: {
        // Transform value to percentage string
        percentage (value) {
            return `${value.toFixed(2)}%`
        },
    },
    mixins: [
        activeMixin,
        themeMixin,
        animationMixin,
    ],
    props: {
        dataset: VueTypes.arrayOf(VueTypes.number).def([]),
        markers: VueTypes.arrayOf(VueTypes.number).def([]),
        trigger: VueTypes.oneOf(['hover', 'click', 'manual']).def('hover'),
        total: VueTypes.number,
        showLabel: VueTypes.bool.def(false),
        delay: VueTypes.number.def(300),
        minWidth: VueTypes.number,
        // Styles
        styles: VueTypes.object.def({}),
        labelStyles: VueTypes.object,
        markerStyles: VueTypes.object.def({}),
    },
    data () {
        return {
            launchAnimation: false,
            offset: this.minWidth || MIN_WIDTH,
        }
    },
    computed: {
        // Used total prop binding or the sum of all values
        totalValues () {
            return this.total || this.sumValues
        },
        // Sum values
        sumValues () {
            return this.dataset.reduce((a, b) => a + b)
        },
        // Generate and calc stack values
        stacks () {
            const values = this.launchAnimation ? this.dataset : Array.from({ length: (this.dataset || []).length })
            return values.reduce((acc, value, index) => {
                if (value === 0) return acc
                const width = (value * 100 / this.totalValues)
                acc.push({
                    id: index,
                    value,
                    color: this.colors[index],
                    width,
                    hide: width < this.offset,
                })
                return acc
            }, [])
        },
        // Generate and calc markers values
        stackMarkers () {
            const markers = this.launchAnimation ? this.markers : []
            return markers.map((value, index) => ({
                id: index,
                left: (value * 100 / this.totalValues),
                value,
            }))
        },
        // Event Listeners
        stackListeners () {
            return merge({}, this.$listeners, {
                click: (event) => {
                    if (this.trigger === 'click') this.handleActive(event)
                    this.$emit('onClick', this.info)
                },
                mouseenter: (event) => {
                    if (this.trigger === 'hover') this.handleActive(event)
                    this.$emit('onMouseenter')
                },
                mouseleave: () => {
                    if (['hover', 'click'].includes(this.trigger)) this.cleanActive()
                    this.$emit('onMouseleave')
                },
            })
        },
        markerListeners () {
            return merge({}, this.$listeners, {
                click: (event) => {
                    if (this.trigger === 'click') this.handleActiveMarker(event)
                    this.$emit('onClickMarker', this.info)
                },
                mouseenter: (event) => {
                    if (this.trigger === 'hover') this.handleActiveMarker(event)
                    this.$emit('onMouseenterMarker')
                },
                mouseleave: () => {
                    if (['hover', 'click'].includes(this.trigger)) this.cleanActive()
                    this.$emit('onMouseleaveMarker')
                },
            })
        },
        // Styles
        barStylesCmp () {
            return {
                ...this.themeStyles.styles,
                ...this.styles,
                ...(this.trigger === 'click' ? { cursor: 'pointer' } : {}),
            }
        },
        labelStylesCmp () {
            return {
                ...this.themeStyles.label,
                ...this.labelStyles,
            }
        },
        markerStylesCmp () {
            return {
                ...this.themeStyles.marker,
                ...this.markerStyles,
                ...(this.trigger === 'click' ? { cursor: 'pointer' } : {}),
            }
        },
    },
    mounted () {
        // Added delay to animate stack
        setTimeout(this.launch, this.delay)
    },
    methods: {
        // Enable launch animation
        launch () {
            this.launchAnimation = true
        },
        // Set active marker, to show tooltip
        handleActiveMarker (event) {
            const { id, offsetLeft } = event.target
            const value = [{ key: id, color: this.markerStylesCmp.background, value: this.stackMarkers[id].value }]
            this.setActive({
                id, value, offsetLeft,
            }, event)
        },
        // Set active bar, to show tooltip
        handleActive (event, isMarker) {
            const { id, offsetLeft } = event.target
            const data = this.stacks.find(stack => stack.id === parseInt(id, 0)) || {}
            const value = [{
                key: id,
                color: this.colors[id],
                value: data.value,
                width: data.width,
            }]
            return this.setActive({
                id, value, data, offsetLeft,
            }, event)
        },
    },
}
</script>
<style scoped lang="scss">
.center {
    position: relative;
    display: flex;
    flex: 1;
    align-items: center;
}

.WSimpleHStackBar {
    display: flex;
    justify-content: space-between;
    align-items: center;
    width: 100%;
}

.Container {
    @extend .center
}

.Value {
    max-width: 100%;
    font-size: 12px;
    font-weight: bold;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
}

.Stacks {
    @extend .center;
}

.Stack {
    @extend .center;
    width: 100%;
    max-width: 0%;
    transition: all 250ms ease;

    &.First {
        border-top-left-radius: 3px;
        border-bottom-left-radius: 3px;
    }

    &.Last {
        border-top-right-radius: 3px;
        border-bottom-right-radius: 3px;
    }
}

.Toolip {
    display: flex;
    align-items: center;
}
</style>
