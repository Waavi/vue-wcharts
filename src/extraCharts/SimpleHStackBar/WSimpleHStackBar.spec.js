import { mount, shallowMount } from '@vue/test-utils'
import WSimpleHStackBar from './WSimpleHStackBar.vue'

describe('Components/WSimpleHStackBar', () => {
    const dataset = [37.3, 10.3, 23.6]
    const propsData = {
        total: 71.6,
        dataset,
        data: dataset,
    }

    const defaultConfig = {
        propsData,
    }

    xit(`Should be render correctly`, () => {
        const wrapper = shallowMount(WSimpleHStackBar, defaultConfig)
        wrapper.setData({ launchAnimation: true })
        expect(wrapper.html()).toMatchSnapshot()
    })

    xit(`Should be render correctly without stacks`, () => {
        const customConfig = { ...defaultConfig, propsData: { } }
        const wrapper = shallowMount(WSimpleHStackBar, customConfig)
        expect(wrapper.html()).toMatchSnapshot()
    })

    xit(`Should be render correctly without total`, () => {
        const customConfig = { ...defaultConfig, propsData: { dataset } }
        const wrapper = shallowMount(WSimpleHStackBar, customConfig)
        expect(parseInt(wrapper.vm.totalValues, 10)).toEqual(71)
    })

    xit(`Should be render correctly with labels`, () => {
        const customConfig = { ...defaultConfig, propsData: { ...defaultConfig.propsData, showLabel: true } }
        const wrapper = shallowMount(WSimpleHStackBar, customConfig)
        wrapper.setData({ launchAnimation: true })
        expect(wrapper.html()).toMatchSnapshot()
    })

    xit(`Should be render correctly with markers`, () => {
        const customConfig = { ...defaultConfig, propsData: { ...defaultConfig.propsData, showLabel: true, markers: [10, 25] } }
        const wrapper = shallowMount(WSimpleHStackBar, customConfig)
        wrapper.setData({ launchAnimation: true })
        expect(wrapper.html()).toMatchSnapshot()
    })

    xit(`Should be render correctly with  3 markers`, () => {
        const customConfig = { ...defaultConfig, propsData: { ...defaultConfig.propsData, showLabel: true, markers: [1, 10, 25] } }
        const wrapper = shallowMount(WSimpleHStackBar, customConfig)
        wrapper.setData({ launchAnimation: true })
        expect(wrapper.findAll('.Marker')).toHaveLength(3)
    })

    xit(`Should be render correctly an invisible marker`, () => {
        const customConfig = { ...defaultConfig, propsData: { ...defaultConfig.propsData, showLabel: true, markers: [-10] } }
        const wrapper = shallowMount(WSimpleHStackBar, customConfig)
        wrapper.setData({ launchAnimation: true })
        expect(wrapper.find('.Marker').isVisible()).toBe(false)
    })

    describe('Events stack', () => {
        xit(`It emits the onClick event`, () => {
            const wrapper = mount(WSimpleHStackBar, defaultConfig)
            wrapper.setData({ launchAnimation: true })
            wrapper.find('.Stacks > .Stack').trigger('click')
            expect(wrapper.emitted('onClick')).toHaveLength(1)
        })

        xit(`It emits the onMouseenter event`, () => {
            const wrapper = mount(WSimpleHStackBar, {
                ...defaultConfig,
                propsData: {
                    ...defaultConfig.propsData,
                    trigger: 'manual',
                },
            })
            wrapper.setData({ launchAnimation: true })
            wrapper.find('.Stacks > .Stack').trigger('mouseenter')
            expect(wrapper.emitted('onMouseenter')).toHaveLength(1)
        })

        xit(`It emits the onMouseleave event`, () => {
            const wrapper = mount(WSimpleHStackBar, {
                ...defaultConfig,
                propsData: {
                    ...defaultConfig.propsData,
                    trigger: 'hover',
                },
            })
            wrapper.setData({ launchAnimation: true })
            wrapper.find('.Stacks > .Stack').trigger('mouseleave')
            expect(wrapper.emitted('onMouseleave')).toHaveLength(1)
        })
    })

    describe('marker stack', () => {
        xit(`It emits the onClick event`, () => {
            const wrapper = mount(WSimpleHStackBar, {
                propsData: {
                    ...defaultConfig.propsData,
                    trigger: 'click',
                    showLabel: true,
                    markers: [10, 25],
                },
            })
            wrapper.setData({ launchAnimation: true })
            wrapper.find('.Marker').trigger('click')
            expect(wrapper.emitted('onClickMarker')).toHaveLength(1)
        })

        xit(`It emits the onMouseenter event`, () => {
            const wrapper = mount(WSimpleHStackBar, {
                ...defaultConfig,
                propsData: {
                    ...defaultConfig.propsData,
                    showLabel: true,
                    markers: [10, 25],
                    trigger: 'manual',
                },
            })
            wrapper.setData({ launchAnimation: true })
            wrapper.find('.Marker').trigger('mouseenter')
            expect(wrapper.emitted('onMouseenterMarker')).toHaveLength(1)
        })

        xit(`It emits the onMouseleave event`, () => {
            const wrapper = mount(WSimpleHStackBar, {
                ...defaultConfig,
                propsData: {
                    ...defaultConfig.propsData,
                    showLabel: true,
                    markers: [10, 25],
                    trigger: 'hover',
                },
            })
            wrapper.setData({ launchAnimation: true })
            wrapper.find('.Marker').trigger('mouseleave')
            expect(wrapper.emitted('onMouseleaveMarker')).toHaveLength(1)
        })
    })

    describe('Methods', () => {
        xit(`Should be set launchAnimation data on call launch method`, () => {
            const wrapper = shallowMount(WSimpleHStackBar, defaultConfig)
            wrapper.vm.launch()
            expect(wrapper.vm.launchAnimation).toBeTruthy()
        })

        xit(`Should be set active in Cahrt after trigger event`, (done) => {
            const setActive = ({
                id, value, datum, offsetLeft,
            }) => {
                expect(id).toEqual(0)
                expect(value).toEqual([{
                    color: '#48c0b6', key: 0, value: 37.3, width: 52.0949720670391,
                }])
                expect(datum).toEqual({
                    color: '#48c0b6', hide: false, id: 0, value: 37.3, width: 52.0949720670391,
                })
                expect(offsetLeft).toEqual(10)
                done()
            }
            const wrapper = mount(WSimpleHStackBar, defaultConfig)
            wrapper.setData({ launchAnimation: true })
            wrapper.vm.setActive = setActive
            const event = {
                target: {
                    id: 0,
                    offsetLeft: 10,
                },
            }
            wrapper.vm.handleActive(event)
        })

        xit(`Should be set active in Cahrt after trigger event without stacks`, (done) => {
            const setActive = ({
                id, value, datum, offsetLeft,
            }) => {
                expect(id).toEqual(0)
                expect(value).toEqual([{
                    color: '#48c0b6', key: 0, value: undefined, width: undefined,
                }])
                expect(datum).toEqual({})
                expect(offsetLeft).toEqual(10)
                done()
            }
            const wrapper = mount(WSimpleHStackBar, {
                ...defaultConfig,
                propsData: {
                    ...propsData,
                    dataset: [],
                },
            })
            wrapper.setData({ launchAnimation: true })
            wrapper.vm.setActive = setActive
            const event = {
                target: {
                    id: 0,
                    offsetLeft: 10,
                },
            }
            wrapper.vm.handleActive(event)
        })
    })
})
