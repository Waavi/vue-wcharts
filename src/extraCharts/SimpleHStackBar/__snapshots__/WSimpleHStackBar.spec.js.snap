// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/WLine Should be render correctly 1`] = `
<div class="WSimpleHStackBar" data="37.3,10.3,23.6">
  <div class="Container">
    <div class="Start"></div>
    <div class="Stacks">
      <div id="0" class="Stack First" style="opacity: 1; background: rgb(72, 192, 182); max-width: 52.0949720670391%;">
        <div class="Value">
          <wshowiffit-stub>
            <!---->
          </wshowiffit-stub>
        </div>
      </div>
      <div id="1" class="Stack" style="opacity: 1; background: rgb(157, 106, 245); max-width: 14.385474860335197%;">
        <div class="Value">
          <wshowiffit-stub>
            <!---->
          </wshowiffit-stub>
        </div>
      </div>
      <div id="2" class="Stack Last" style="opacity: 1; background: rgb(255, 222, 105); max-width: 32.960893854748605%;">
        <div class="Value">
          <wshowiffit-stub>
            <!---->
          </wshowiffit-stub>
        </div>
      </div>
    </div>
    <wtooltip-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" id="" gap="10" styles="[object Object]" visiblestyles="[object Object]" titlestyles="[object Object]" wrapperstyles="[object Object]"></wtooltip-stub>
    <div class="End"></div>
  </div>
</div>
`;

exports[`Components/WLine Should be render correctly with labels 1`] = `
<div class="WSimpleHStackBar" data="37.3,10.3,23.6">
  <div class="Container">
    <div class="Start"></div>
    <div class="Stacks">
      <div id="0" class="Stack First" style="opacity: 1; background: rgb(72, 192, 182); max-width: 52.0949720670391%;">
        <div class="Value">
          <wshowiffit-stub><span style="color: rgb(72, 192, 182);">37.3</span></wshowiffit-stub>
        </div>
      </div>
      <div id="1" class="Stack" style="opacity: 1; background: rgb(157, 106, 245); max-width: 14.385474860335197%;">
        <div class="Value">
          <wshowiffit-stub><span style="color: rgb(157, 106, 245);">10.3</span></wshowiffit-stub>
        </div>
      </div>
      <div id="2" class="Stack Last" style="opacity: 1; background: rgb(255, 222, 105); max-width: 32.960893854748605%;">
        <div class="Value">
          <wshowiffit-stub><span style="color: rgb(255, 222, 105);">23.6</span></wshowiffit-stub>
        </div>
      </div>
    </div>
    <wtooltip-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" id="" gap="10" styles="[object Object]" visiblestyles="[object Object]" titlestyles="[object Object]" wrapperstyles="[object Object]"></wtooltip-stub>
    <div class="End"></div>
  </div>
</div>
`;

exports[`Components/WLine Should be render correctly with markers 1`] = `
<div class="WSimpleHStackBar" data="37.3,10.3,23.6">
  <div class="Container">
    <div class="Start"></div>
    <div class="Stacks">
      <div id="0" class="Stack First" style="opacity: 1; background: rgb(72, 192, 182); max-width: 52.0949720670391%;">
        <div class="Value">
          <wshowiffit-stub><span style="color: rgb(72, 192, 182);">37.3</span></wshowiffit-stub>
        </div>
      </div>
      <div id="1" class="Stack" style="opacity: 1; background: rgb(157, 106, 245); max-width: 14.385474860335197%;">
        <div class="Value">
          <wshowiffit-stub><span style="color: rgb(157, 106, 245);">10.3</span></wshowiffit-stub>
        </div>
      </div>
      <div id="2" class="Stack Last" style="opacity: 1; background: rgb(255, 222, 105); max-width: 32.960893854748605%;">
        <div class="Value">
          <wshowiffit-stub><span style="color: rgb(255, 222, 105);">23.6</span></wshowiffit-stub>
        </div>
      </div>
    </div>
    <wtooltip-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" id="" gap="10" styles="[object Object]" visiblestyles="[object Object]" titlestyles="[object Object]" wrapperstyles="[object Object]"></wtooltip-stub>
    <div id="0" data-type="marker" class="Marker" style="left: 13.966480446927376%;">
      <div data-value="10"></div>
    </div>
    <div id="1" data-type="marker" class="Marker" style="left: 34.91620111731844%;">
      <div data-value="25"></div>
    </div>
    <div class="End"></div>
  </div>
</div>
`;

exports[`Components/WLine Should be render correctly without stacks 1`] = `
<div class="WSimpleHStackBar">
  <div class="Container">
    <!---->
    <div class="Stacks"></div>
    <wtooltip-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" id="" gap="10" styles="[object Object]" visiblestyles="[object Object]" titlestyles="[object Object]" wrapperstyles="[object Object]"></wtooltip-stub>
    <!---->
  </div>
</div>
`;
